pkg/osm/osm_controller_logs.go

```diff
@@ -0,0 +1,67 @@
+package osm
+
+import (
+	"context"
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/openservicemesh/osm-health/pkg/common"
+	"github.com/openservicemesh/osm-health/pkg/kubernetes/podhelper"
+	"github.com/openservicemesh/osm/pkg/constants"
+)
+
+// Verify interface compliance
+var _ common.Runnable = (*NoBadOsmControllerLogsCheck)(nil)
+
+// NoBadOsmControllerLogsCheck implements common.Runnable
+type NoBadOsmControllerLogsCheck struct {
+	client                   kubernetes.Interface
+	osmControlPlaneNamespace string
+}
+
+// HasNoBadOsmControllerLogsCheck checks whether the osm controller pods in the controller namespace have bad (fatal/error/warning/fail) log messages
+func HasNoBadOsmControllerLogsCheck(client kubernetes.Interface, osmControlPlaneNamespace string) NoBadOsmControllerLogsCheck {
+	return NoBadOsmControllerLogsCheck{
+		client:                   client,
+		osmControlPlaneNamespace: osmControlPlaneNamespace,
+	}
+}
+
+// Info implements common.Runnable
+func (check NoBadOsmControllerLogsCheck) Info() string {
+	return fmt.Sprintf("Checking whether namespace %s has bad (fatal/error/warning/fail) logs in osm controller pods", check.osmControlPlaneNamespace)
+}
+
+// Run implements common.Runnable
+func (check NoBadOsmControllerLogsCheck) Run() error {
+	labelSelector := metav1.LabelSelector{MatchLabels: map[string]string{"app": constants.OSMControllerName}}
+	listOptions := metav1.ListOptions{
+		LabelSelector: labels.Set(labelSelector.MatchLabels).String(),
+	}
+	pods, err := check.client.CoreV1().Pods(check.osmControlPlaneNamespace).List(context.TODO(), listOptions)
+	if err != nil {
+		return fmt.Errorf("unable to list osm controller pods in namespace %s", check.osmControlPlaneNamespace)
+	}
+
+	for i := range pods.Items {
+		if err := podhelper.HasNoBadLogs(check.client, &pods.Items[i], "osm-controller"); err != nil {
+			return err // TODO since we can have multiple osm-controller pods, should we return err on the first controller with bad logs?
```

Comment:
Would love your thoughts on this.

Reply:
option 1: would it be possible to iterate over all the controller pods and run this check on each of them from wherever this is being called (similar to what we did for aks-periscope)? Because I would imagine eventually we will have other checks on the controller pods too so this scenario is going to arise for each of them so it would be more comprehensive to do that. If not, then yes I think it would make sense to return err on the first one and hope users will run checks again after debugging the first one

option 2: have them specify the meshName or controller pod name in the command but I feel like this is less ideal

Reply:
Refactored error handling and logging so that all osm-controller log errors are exposed. Done through this commit: https://github.com/openservicemesh/osm-health/pull/50/commits/be6bea974439042d3e7ab8d2283f8366bbb5ce31